### PR TITLE
Kdesktop 783 implement a new table key value in parms db

### PR DESCRIPTION
**Pull request by herve-er**
This new table will allow us to store value under a key and to retrieve it.
It can be useful in different situation:
- Save the last time of serverSelfRestart.
- idem for the client.
- save the timestamp of the last successful log upload.
- save the path of the log archive in case of upload fail
- ….

A request will also be set up to allow the client to set/retrieve data from this table.

### CODE REVIEW

File: src/libcommon/comm.h [hunk 1]
```diff
@@ -109,6 +109,8 @@ typedef enum {
     REQUEST_NUM_UTILITY_CHECKCOMMSTATUS,
     REQUEST_NUM_UTILITY_HASSYSTEMLAUNCHONSTARTUP,
     REQUEST_NUM_UTILITY_HASLAUNCHONSTARTUP,
+    REQUEST_NUM_UTILITY_SET_KEYVALUE,
+    REQUEST_NUM_UTILITY_GET_KEYVALUE,
     REQUEST_NUM_UTILITY_SETLAUNCHONSTARTUP,
```
Comment (ChristopheLarchier) [inline src/libcommon/comm.h:114]: To move after REQUEST_NUM_UTILITY_HASLAUNCHONSTARTUP

File: src/libparms/db/parmsdb.cpp [hunk 2]
```diff
@@ -3371,4 +3445,77 @@ bool ParmsDb::updateLastClientSelfRestartTime(int64_t lastClientRestartTime) {
     return true;
 }
 
+
+bool ParmsDb::selectValueForKey(const std::string &key, std::string &value, const std::string &defaultValue) {
```
Comment (ChristopheLarchier) [inline src/libparms/db/parmsdb.cpp:3449]: This kind of functions usually has a "bool &found" parameter and no defaultValue

File: src/libparms/db/parmsdb.cpp [hunk 3]
```diff
@@ -3371,4 +3445,77 @@ bool ParmsDb::updateLastClientSelfRestartTime(int64_t lastClientRestartTime) {
...
+        if (!queryExec(INSERT_KEY_VALUE_REQUEST_ID, errId, error)) {
+            LOG_WARN(_logger, "Error running query: " << INSERT_KEY_VALUE_REQUEST_ID);
+            return false;
+        }
+        ASSERT(queryResetAndClearBindings(INSERT_KEY_VALUE_REQUEST_ID));
+    }
+    return true;
+}
+
+bool ParmsDb::deleteKeyValueEntry(const std::string &key) {
```
Comment (ChristopheLarchier) [inline src/libparms/db/parmsdb.cpp:3505]: Should not exist.

File: src/libcommon/comm.h (hunk 1 above)
Comment (herve-er) [reply src/libcommon/comm.h]: I movedit in commit: https://github.com/Infomaniak/desktop-kDrive/pull/76/commits/0a3ea500340aec701b225bf3ee1c689a645865f5

File: src/libparms/db/parmsdb.cpp (hunk 3 above)
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: As suggested, I removed this function (https://github.com/Infomaniak/desktop-kDrive/pull/76/commits/2947e9f121aeeef3d61871f1601dd6cda6d61ee3). The only case I found where we might want to permanently delete an entry is during an app update when we no longer use a keyValue. In this scenario, there's nothing preventing us from implementing a similar function at that time if there are a lot of unused keyValues.

File: src/libparms/db/parmsdb.cpp (hunk 2 above)
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp:3440]: I think using defaultValue can make the caller code easier to read. For exemple:
```
bool found = false;
std::string value;

if(!selectValueForKey("lasLogUpload", value, found)){
    //handle the error
}

if(!found){
   //hanlde the not found
}
// use of the value
```

become:
```
std::string value;

if(!selectValueForKey("lasLogUpload", value, "01/01/2000_00h00m00s")){
    //handle the error
}
// use of the value
```

or:

```
std::string value;
selectValueForKey("lasLogUpload", value, "01/01/2000_00h00m00s")){
// use of the value
```

But I also understand the benefice of keeping the same mechanism across the function when possible

File: src/gui/guirequests.cpp [hunk 4]
```diff
@@ -823,6 +823,44 @@ ExitCode GuiRequests::setLaunchOnStartup(bool enabled) {
...
+    QDataStream paramsStream(&params, QIODevice::WriteOnly);
+    paramsStream << key;
+    paramsStream << defaultValue;
+
+    QByteArray results;
+    if (!CommClient::instance()->execute(REQUEST_NUM_UTILITY_GET_KEYVALUE, params, results)) {
+        return ExitCodeSystemError;
+    }
+
+    ExitCode exitCode;
```
Comment (luc-guyot-infomaniak) [inline src/gui/guirequests.cpp:837]:
```suggestion
    ExitCode exitCode = ExitCodeUnknown;
```

File: src/gui/guirequests.cpp [hunk 5]
```diff
@@ -823,6 +823,44 @@ ExitCode GuiRequests::setLaunchOnStartup(bool enabled) {
...
+    QDataStream paramsStream(&params, QIODevice::WriteOnly);
+    paramsStream << key;
+    paramsStream << value;
+
+    QByteArray results;
+    if (!CommClient::instance()->execute(REQUEST_NUM_UTILITY_SET_KEYVALUE, params, results)) {
+        return ExitCodeSystemError;
+    }
+
+    ExitCode exitCode;
```
Comment (luc-guyot-infomaniak) [inline src/gui/guirequests.cpp:856]:
```suggestion
    ExitCode exitCode = ExitCodeUnknown;
```

File: src/gui/guirequests.cpp [hunk 6]
```diff
@@ -823,6 +823,44 @@ ExitCode GuiRequests::setLaunchOnStartup(bool enabled) {
...
+    QDataStream resultStream(&results, QIODevice::ReadOnly);
+    resultStream >> exitCode;
+
+    return exitCode;
+
```
Comment (luc-guyot-infomaniak) [inline src/gui/guirequests.cpp:861]: You may remove this empty line.

File: src/libparms/db/parmsdb.cpp [hunk 7]
```diff
@@ -525,6 +525,27 @@
 #define UPDATE_SELF_RESTARTER_SERVER_REQUEST_ID "update_self_restarter_server_time"
 #define UPDATE_SELF_RESTARTER_SERVER_REQUEST "UPDATE self_restarter SET lastServerRestart=?1;"
 
+//
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdb.cpp:528]: This file is huge. Would it be possible to create a new header and a new cpp file to introduce the functions that would be called here?

File: src/libparms/db/parmsdb.cpp [hunk 8]
```diff
@@ -3371,4 +3436,55 @@ bool ParmsDb::updateLastClientSelfRestartTime(int64_t lastClientRestartTime) {
...
+
+    ASSERT(queryStringValue(SELECT_VALUE_FROM_KEY_REQUEST_ID, 0, value));
+    ASSERT(queryResetAndClearBindings(SELECT_VALUE_FROM_KEY_REQUEST_ID));
+
+    return true;
+}
+
+bool ParmsDb::setValueForKey(const std::string &key, const std::string &value) {
+    std::string existingValue;
+    int errId;
```
Comment (luc-guyot-infomaniak) [inline src/libparms/db/parmsdb.cpp:3463]:
```suggestion
    int errId = 0;
```

File: src/gui/guirequests.cpp (hunk 5 above)
Comment (herve-er) [reply src/gui/guirequests.cpp]: idem.

File: src/gui/guirequests.cpp (hunk 6 above)
Comment (herve-er) [reply src/gui/guirequests.cpp]: idem.

File: src/libparms/db/parmsdb.cpp (hunk 8 above)
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: I edited it.

File: src/libparms/db/parmsdb.cpp (hunk 7 above)
Comment (herve-er) [reply src/libparms/db/parmsdb.cpp]: I split the code relating to this PR in new `.h` and `.cpp` files. I will create a Jira ticket to plan the same refactor for the rest of the file.

File: test/libparms/db/testparmsdb.cpp [hunk 9]
```diff
@@ -299,6 +298,20 @@ void TestParmsDb::testExclusionTemplate() {
     CPPUNIT_ASSERT(ParmsDb::instance()->deleteExclusionTemplate(exclusionTemplate3.templ(), found) && found);
 }
 
+void TestParmsDb::testAppState(void) {
+    bool found = true;
+    std::string value;
+    CPPUNIT_ASSERT(ParmsDb::instance()->selectAppState(AppStateKeyTest, value, found) && found);
+    CPPUNIT_ASSERT(value == "Test");
```
Comment (luc-guyot-infomaniak) [inline test/libparms/db/testparmsdb.cpp:305]:
```suggestion
    CPPUNIT_ASSERT_EQUAL("Test", value);
```

File: test/libparms/db/testparmsdb.cpp [hunk 10]
```diff
@@ -299,6 +298,20 @@ void TestParmsDb::testExclusionTemplate() {
...
+
+    CPPUNIT_ASSERT(ParmsDb::instance()->updateAppState(AppStateKeyTest, "value 1", found));
+    CPPUNIT_ASSERT(ParmsDb::instance()->selectAppState(AppStateKeyTest, value, found) && found);
+    CPPUNIT_ASSERT(value == "value 1");
```
Comment (luc-guyot-infomaniak) [inline test/libparms/db/testparmsdb.cpp:309]:
```suggestion
    CPPUNIT_ASSERT_EQUAL("value 1", value);
```

File: src/libcommon/utility/types.h [hunk 11]
```diff
@@ -358,4 +358,10 @@ struct ItemType {
         // `LinkTypeUnknown`) and its target doesn't exist.
         IoError ioError{IoErrorSuccess};
 };
+
+typedef enum {
```
Comment (luc-guyot-infomaniak) [inline src/libcommon/utility/types.h:362]: Use `enum AppStateKey ` instead (see https://en.cppreference.com/w/cpp/language/enum). Sonar Source should give hints as how to proceed in the best way,

**Review by luc-guyot-infomaniak — APPROVED** [review]

File: src/libcommon/utility/types.h (hunk 11 above)
Comment (herve-er) [reply src/libcommon/utility/types.h]: Following the Sonar advice (and my programming course) it is even better to use an enum class. (commit: https://github.com/Infomaniak/desktop-kDrive/pull/76/commits/3b8cefecd5477d510239ad455ce38b1ced471f34)

**Review by ChristopheLarchier — APPROVED** [review]

**Pull request merged**